builder/dockerfile/dispatchers.go

```diff
@@ -208,65 +141,76 @@ func (b *Builder) getImageMount(fromFlag *Flag) (*imageMount, error) {
 
 // FROM imagename[:tag | @digest] [AS build-stage-name]
 //
-func from(req dispatchRequest) error {
-	stageName, err := parseBuildStageName(req.args)
-	if err != nil {
-		return err
-	}
-
-	if err := req.flags.Parse(); err != nil {
-		return err
-	}
-
-	req.builder.imageProber.Reset()
-	image, err := req.builder.getFromImage(req.shlex, req.args[0])
+func dispatchFrom(d *stageBuild, cmd *instructions.FromCommand) error {
+	d.builder.imageProber.Reset()
+	image, err := d.getFromImage(d.shlex, cmd.BaseName)
 	if err != nil {
 		return err
 	}
-	if err := req.builder.buildStages.add(stageName, image); err != nil {
+	if err := d.builder.buildStages.add(cmd.StageName, image); err != nil {
 		return err
 	}
-	req.state.beginStage(stageName, image)
-	req.builder.buildArgs.ResetAllowed()
-	if image.ImageID() == "" {
-		// Typically this means they used "FROM scratch"
-		return nil
+	state := d.state
+	state.beginStage(cmd.StageName, image)
+	if state.runConfig.OnBuild != nil && len(state.runConfig.OnBuild) > 0 {
```

Comment:
This first nil check is unnecessary, you can just check length (https://play.golang.org/p/E0QPccyYAU)